File: apps/api/src/services/model/router.ts
// c03｜用途路由 + 优先级 fallback + 公网前置脱敏门控 + 出网网关编排（design D2/D4/D6）
// 唯一对上层暴露的调用入口：invokeChat/invokeTranslate/.../invokeEmbed/invokeRerank。
// 上层不感知协议/provider/公网私有化；fallback、脱敏门禁、切换均落 audit_logs。
import type { PoolClient } from "pg";
import { writeAudit } from "../audit.js";
import type {
  Capability,
  EmbedRequest,
  EmbedResponse,
  GenerationRequest,
  GenerationResponse,
  RerankRequest,
  RerankResponse,
} from "./capabilities.js";
import { getAdapter, type ModelAdapter } from "./adapters.js";
import { isFallbackable, ProviderError, type ProviderConnection } from "./http.js";
import { redactionGateway } from "./redaction-gateway.js";
import { recordHealth, resolveModelChain } from "./provider-store.js";

export interface InvokeContext {
  tenantId: string;
  actorId?: string | null;
  actorRole?: string | null;
}

export class CapabilityUnavailableError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "CapabilityUnavailableError";
  }
}

interface ChainCall<T> {
  capability: Capability;
  /** 公网外发前交 c09 redaction-gateway 判定的原文 */
  outboundText: string;
  run(adapter: ModelAdapter, conn: ProviderConnection): Promise<T>;
}

async function runChain<T>(
  client: PoolClient,
  ctx: InvokeContext,
  call: ChainCall<T>,
): Promise<T> {
  const chain = await resolveModelChain(client, ctx.tenantId, call.capability);
  if (!chain.length) {
    await writeAudit(client, {
      tenantId: ctx.tenantId,
      actorId: ctx.actorId,
      actorRole: ctx.actorRole,
      actionType: "model_route_missing",
      targetType: "capability",
      targetId: call.capability,
      result: "失败",
      failureReason: "该用途未配置可用模型",
    });
    throw new CapabilityUnavailableError(`该用途未配置可用模型：${call.capability}`);
  }

  for (let i = 0; i < chain.length; i++) {
    const conn = chain[i];
    const next = chain[i + 1] ?? null;

    // D6：公网 provider 必须先过 c09 脱敏门禁；未通过则跳过公网，尝试下一（私有化）provider
    if (conn.deploymentKind === "public") {
      const verdict = await redactionGateway.evaluate({
        tenantId: ctx.tenantId,
        text: call.outboundText,
      });
      if (!verdict.available || !verdict.passed) {
        await writeAudit(client, {
          tenantId: ctx.tenantId,
          actorId: ctx.actorId,
          actorRole: ctx.actorRole,
          actionType: "model_redaction_block",
          targetType: "capability",
          targetId: call.capability,
          result: "失败",
          failureReason: verdict.reason,
          metadata: {
            provider: conn.name,
            providerId: conn.providerId,
            deploymentKind: "public",
            switchTo: next?.name ?? null,
            confidence: verdict.confidence,
          },
        });
        continue; // 跳过公网，绝不在未脱敏情况下外发
      }
      // c09 落地后：verdict.redactedText 即脱敏后内容，应替换 payload 再外发（本期不可达）
    }

    const adapter = getAdapter(conn.protocol ?? "");
    let lastError: ProviderError | null = null;
    for (let attempt = 0; attempt <= conn.maxRetries; attempt++) {
      try {
        const result = await call.run(adapter, conn);
        await writeAudit(client, {
          tenantId: ctx.tenantId,
          actorId: ctx.actorId,
          actorRole: ctx.actorRole,
          actionType: "model_invoke",
          targetType: "capability",
          targetId: call.capability,
          result: "成功",
          metadata: {
            provider: conn.name,
            providerId: conn.providerId,
            deploymentKind: conn.deploymentKind,
            attempt,
          },
        });
        return result;
      } catch (e) {
        const err =
          e instanceof ProviderError ? e : new ProviderError("unknown", (e as Error).message);
        lastError = err;
        if (!isFallbackable(err.errorClass)) {
          // 不可 fallback（鉴权错/超上限/内容安全）→ 直接终止链路上抛
          await writeAudit(client, {
            tenantId: ctx.tenantId,
            actorId: ctx.actorId,
            actorRole: ctx.actorRole,
            actionType: "model_invoke",
            targetType: "capability",
            targetId: call.capability,
            result: "失败",
            failureReason: err.message,
            metadata: { provider: conn.name, providerId: conn.providerId, errorClass: err.errorClass },
          });
          throw err;
        }
        // 可 fallback：继续该 provider 的重试，耗尽后切换
      }
    }

    // 该 provider 重试耗尽 → 被动标记 down + 记录 fallback 切换（四要素：provider/失败原因/切换目标/时间戳）
    await recordHealth(client, {
      tenantId: ctx.tenantId,
      providerId: conn.providerId,
      providerKind: "model",
      checkKind: "passive",
      status: "down",
      error: lastError?.message ?? "调用失败",
    });
    await writeAudit(client, {
      tenantId: ctx.tenantId,
      actorId: ctx.actorId,
      actorRole: ctx.actorRole,
      actionType: "model_fallback",
      targetType: "capability",
      targetId: call.capability,
      result: "失败",
      failureReason: lastError?.message ?? "调用失败",
      metadata: {
        fromProvider: conn.name,
        fromProviderId: conn.providerId,
        errorClass: lastError?.errorClass ?? "unknown",
        toProvider: next?.name ?? null,
        toProviderId: next?.providerId ?? null,
        capability: call.capability,
      },
    });
  }

  // 全链路失败/被跳过 → 明确不可用错误（绝不静默成功）
  await writeAudit(client, {
    tenantId: ctx.tenantId,
    actorId: ctx.actorId,
    actorRole: ctx.actorRole,
    actionType: "model_unavailable",
    targetType: "capability",
    targetId: call.capability,
    result: "失败",
    failureReason: "该用途所有 provider 依次失败或被脱敏门禁拒绝",
  });
  throw new CapabilityUnavailableError(
    `该用途所有 provider 均不可用：${call.capability}`,
  );
}

// ——— 对上层暴露的 capability 调用入口 ———

export function invokeGeneration(
  client: PoolClient,
  capability: Capability,
  req: GenerationRequest,
  ctx: InvokeContext,
): Promise<GenerationResponse> {
  return runChain(client, ctx, {
    capability,
    outboundText: req.messages.map((m) => m.content).join("\n"),
    run: (adapter, conn) => adapter.generate(req, conn),
  });
}

export function invokeEmbed(
  client: PoolClient,
  req: EmbedRequest,
  ctx: InvokeContext,
): Promise<EmbedResponse> {
  return runChain(client, ctx, {
    capability: "embed",
    outboundText: req.input.join("\n"),
    run: (adapter, conn) => adapter.embed(req, conn),
  });
}

export function invokeRerank(
  client: PoolClient,
  req: RerankRequest,
  ctx: InvokeContext,
): Promise<RerankResponse> {
  return runChain(client, ctx, {
    capability: "rerank",
    outboundText: [req.query, ...req.documents].join("\n"),
    run: (adapter, conn) => adapter.rerank(req, conn),
  });
}

File: apps/server/internal/model/router.go
package model

import (
	"errors"
	"fmt"

	"gorm.io/gorm"

	"medoffice/server/internal/audit"
)

type InvokeContext struct {
	TenantID  string
	ActorID   string
	ActorRole string
}

type CapabilityUnavailableError struct{ Msg string }

func (e *CapabilityUnavailableError) Error() string { return e.Msg }

func asProviderError(err error) *ProviderError {
	var pe *ProviderError
	if errors.As(err, &pe) {
		return pe
	}
	return newProviderError(ErrUnknown, err.Error())
}

func actorPtr(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}

type chainCall[T any] struct {
	Capability   Capability
	OutboundText string
	Run          func(a Adapter, conn ProviderConnection) (T, error)
}

// runChain 复刻 router.ts runChain：fallback 链 + 公网脱敏门控 + 出网 + 全程审计。
func runChain[T any](db *gorm.DB, ctx InvokeContext, call chainCall[T]) (T, error) {
	var zero T
	chain, err := ResolveModelChain(db, ctx.TenantID, call.Capability)
	if err != nil {
		return zero, err
	}
	if len(chain) == 0 {
		_ = audit.Write(db, audit.Entry{
			TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
			ActionType: "model_route_missing", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
			Result: "失败", FailureReason: audit.P("该用途未配置可用模型"),
		})
		return zero, &CapabilityUnavailableError{fmt.Sprintf("该用途未配置可用模型：%s", string(call.Capability))}
	}

	for i := range chain {
		conn := chain[i]
		var next *ProviderConnection
		if i+1 < len(chain) {
			next = &chain[i+1]
		}
		nextName := func() any {
			if next != nil {
				return next.Name
			}
			return nil
		}

		// D6：公网 provider 必须先过 c09 脱敏门禁；未通过则跳过公网
		if conn.DeploymentKind == DeployPublic {
			v := redactionGateway.Evaluate(RedactionInput{TenantID: ctx.TenantID, Text: call.OutboundText})
			if !v.Available || !v.Passed {
				_ = audit.Write(db, audit.Entry{
					TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
					ActionType: "model_redaction_block", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
					Result: "失败", FailureReason: audit.P(v.Reason),
					Metadata: map[string]any{"provider": conn.Name, "providerId": conn.ProviderID, "deploymentKind": "public", "switchTo": nextName(), "confidence": v.Confidence},
				})
				continue
			}
		}

		adapter, aerr := getAdapter(conn.Protocol)
		if aerr != nil {
			return zero, aerr
		}
		var lastErr *ProviderError
		for attempt := 0; attempt <= conn.MaxRetries; attempt++ {
			result, rerr := call.Run(adapter, conn)
			if rerr == nil {
				_ = audit.Write(db, audit.Entry{
					TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
					ActionType: "model_invoke", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
					Result: "成功", Metadata: map[string]any{"provider": conn.Name, "providerId": conn.ProviderID, "deploymentKind": conn.DeploymentKind, "attempt": attempt},
				})
				return result, nil
			}
			pe := asProviderError(rerr)
			lastErr = pe
			if !IsFallbackable(pe.Class) {
				_ = audit.Write(db, audit.Entry{
					TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
					ActionType: "model_invoke", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
					Result: "失败", FailureReason: audit.P(pe.Msg),
					Metadata: map[string]any{"provider": conn.Name, "providerId": conn.ProviderID, "errorClass": pe.Class},
				})
				return zero, pe
			}
		}

		reason := "调用失败"
		var errClass ErrorClass = ErrUnknown
		if lastErr != nil {
			reason = lastErr.Msg
			errClass = lastErr.Class
		}
		_ = RecordHealth(db, HealthRecord{TenantID: ctx.TenantID, ProviderID: conn.ProviderID, ProviderKind: "model", CheckKind: "passive", Status: "down", Error: reason})
		var toProviderID any
		if next != nil {
			toProviderID = next.ProviderID
		}
		_ = audit.Write(db, audit.Entry{
			TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
			ActionType: "model_fallback", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
			Result: "失败", FailureReason: audit.P(reason),
			Metadata: map[string]any{
				"fromProvider": conn.Name, "fromProviderId": conn.ProviderID, "errorClass": errClass,
				"toProvider": nextName(), "toProviderId": toProviderID, "capability": call.Capability,
			},
		})
	}

	_ = audit.Write(db, audit.Entry{
		TenantID: ctx.TenantID, ActorID: actorPtr(ctx.ActorID), ActorRole: actorPtr(ctx.ActorRole),
		ActionType: "model_unavailable", TargetType: audit.P("capability"), TargetID: audit.P(string(call.Capability)),
		Result: "失败", FailureReason: audit.P("该用途所有 provider 依次失败或被脱敏门禁拒绝"),
	})
	return zero, &CapabilityUnavailableError{fmt.Sprintf("该用途所有 provider 均不可用：%s", string(call.Capability))}
}

func InvokeGeneration(db *gorm.DB, capability Capability, req GenerationRequest, ctx InvokeContext) (GenerationResponse, error) {
	var parts []string
	for _, m := range req.Messages {
		parts = append(parts, m.Content)
	}
	return runChain(db, ctx, chainCall[GenerationResponse]{
		Capability: capability, OutboundText: join(parts),
		Run: func(a Adapter, conn ProviderConnection) (GenerationResponse, error) { return a.Generate(req, conn) },
	})
}

func InvokeEmbed(db *gorm.DB, req EmbedRequest, ctx InvokeContext) (EmbedResponse, error) {
	return runChain(db, ctx, chainCall[EmbedResponse]{
		Capability: CapEmbed, OutboundText: join(req.Input),
		Run: func(a Adapter, conn ProviderConnection) (EmbedResponse, error) { return a.Embed(req, conn) },
	})
}

func InvokeRerank(db *gorm.DB, req RerankRequest, ctx InvokeContext) (RerankResponse, error) {
	parts := append([]string{req.Query}, req.Documents...)
	return runChain(db, ctx, chainCall[RerankResponse]{
		Capability: CapRerank, OutboundText: join(parts),
		Run: func(a Adapter, conn ProviderConnection) (RerankResponse, error) { return a.Rerank(req, conn) },
	})
}

func join(parts []string) string {
	out := ""
	for i, p := range parts {
		if i > 0 {
			out += "\n"
		}
		out += p
	}
	return out
}
